File: plugin/vscode/src/test/cli/Executable.test.ts
import * as assert from "node:assert/strict";
import {
  chmodSync,
  createReadStream,
  existsSync,
  mkdirSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import {
  __setOpenReadStreamForTests,
  executablePath,
} from "../../main/cli/Executable";
import { mockExtensionContext } from "../support/MockContext";

suite("Executable", () => {
  const root = join(tmpdir(), `docsig-exe-test-${process.pid}`);
  const extensionPath = join(root, "ext");
  const storagePath = join(root, "storage");
  const bundled = join(extensionPath, "resources", "docsig.pyz");

  suiteSetup(() => {
    mkdirSync(join(extensionPath, "resources"), { recursive: true });
    mkdirSync(storagePath, { recursive: true });
    writeFileSync(bundled, "bundled-cli");
  });

  suiteTeardown(() => {
    rmSync(root, { recursive: true, force: true });
  });

  setup(() => {
    __setOpenReadStreamForTests();
    const cached = join(storagePath, "docsig.pyz");
    if (existsSync(cached)) {
      rmSync(cached, { recursive: true, force: true });
    }
  });

  teardown(() => {
    __setOpenReadStreamForTests();
  });

  test("path extracts bundled cli when cache missing", async () => {
    const context = mockExtensionContext(extensionPath, storagePath);
    const path = await executablePath(context);
    const cached = join(storagePath, "docsig.pyz");

    assert.equal(path, cached);
    assert.equal(existsSync(cached), true);
    assert.ok(readFileSync(cached).length > 0);
  });

  test("path replaces stale cached bundle", async () => {
    const context = mockExtensionContext(extensionPath, storagePath);
    const cached = join(storagePath, "docsig.pyz");
    writeFileSync(cached, "broken");

    const path = await executablePath(context);

    assert.equal(path, cached);
    assert.ok(readFileSync(cached).length > "broken".length);
  });

  test("path throws when bundled cli is missing", async () => {
    const missingRoot = join(tmpdir(), `docsig-exe-missing-${process.pid}`);
    const missingExt = join(missingRoot, "ext");
    const missingStorage = join(missingRoot, "storage");
    mkdirSync(missingExt, { recursive: true });
    mkdirSync(missingStorage, { recursive: true });

    try {
      const context = mockExtensionContext(missingExt, missingStorage);
      await assert.rejects(
        () => executablePath(context),
        (error: Error) => error.message.includes("missing bundled cli"),
      );
    } finally {
      rmSync(missingRoot, { recursive: true, force: true });
    }
  });

  test("path re-extracts when cached digest throws synchronously", async () => {
    const context = mockExtensionContext(extensionPath, storagePath);
    const cached = join(storagePath, "docsig.pyz");
    writeFileSync(cached, "stale");
    __setOpenReadStreamForTests((path, options) => {
      if (path === cached) {
        throw new Error("cannot open cache");
      }
      return createReadStream(path, options);
    });

    await executablePath(context);
    assert.equal(readFileSync(cached).toString(), "bundled-cli");
  });

  test("path re-extracts when cached digest fails", async () => {
    const context = mockExtensionContext(extensionPath, storagePath);
    const cached = join(storagePath, "docsig.pyz");
    writeFileSync(cached, "stale");
    chmodSync(cached, 0o000);

    try {
      await executablePath(context);
      chmodSync(cached, 0o644);
      assert.equal(readFileSync(cached).toString(), "bundled-cli");
    } finally {
      if (existsSync(cached)) {
        chmodSync(cached, 0o644);
      }
    }
  });

  test("path removes temp file when extract fails", async () => {
    const context = mockExtensionContext(extensionPath, storagePath);
    const cached = join(storagePath, "docsig.pyz");
    mkdirSync(cached, { recursive: true });

    try {
      await assert.rejects(() => executablePath(context));
      const temp = join(tmpdir(), `docsig.pyz.${process.pid}.tmp`);
      assert.equal(existsSync(temp), false);
    } finally {
      rmSync(cached, { recursive: true, force: true });
    }
  });

  test("path reuses valid cached bundle", async () => {
    const context = mockExtensionContext(extensionPath, storagePath);
    const first = await executablePath(context);
    const cached = join(storagePath, "docsig.pyz");
    const modified = readFileSync(cached);
    const second = await executablePath(context);

    assert.equal(first, second);
    assert.deepEqual(readFileSync(cached), modified);
  });
});

File: editors/intellij/src/test/kotlin/com/github/jshwi/docsig/cli/ExecutableTest.kt
package com.github.jshwi.docsig.cli

import com.intellij.openapi.application.PathManager
import io.mockk.every
import io.mockk.mockkStatic
import io.mockk.unmockkAll
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Files
import java.nio.file.Path

class ExecutableTest {
    @AfterEach
    fun teardown() {
        unmockkAll()
    }

    @Test
    fun `path extracts bundled cli when cache missing`(
        @TempDir tempDir: Path,
    ) {
        mockSystemPath(tempDir)

        val cached = tempDir.resolve("docsig.pyz")

        assertTrue(Files.notExists(cached))

        val path = Executable.path()

        assertEquals(cached.toString(), path)
        assertTrue(Files.isRegularFile(cached))
        assertTrue(Files.size(cached) > 0L)
    }

    @Test
    fun `path replaces stale cached bundle`(@TempDir tempDir: Path) {
        mockSystemPath(tempDir)

        val cached = tempDir.resolve("docsig.pyz")

        Files.writeString(cached, "broken")

        val path = Executable.path()

        assertEquals(cached.toString(), path)
        assertTrue(Files.size(cached) > "broken".length.toLong())
    }

    @Test
    fun `path reuses valid cached bundle`(@TempDir tempDir: Path) {
        mockSystemPath(tempDir)

        val cached = tempDir.resolve("docsig.pyz")

        val first = Executable.path()
        val modified = Files.getLastModifiedTime(cached)
        val second = Executable.path()

        assertEquals(first, second)
        assertEquals(modified, Files.getLastModifiedTime(cached))
    }

    private fun mockSystemPath(tempDir: Path) {
        mockkStatic(PathManager::class)

        every {
            PathManager.getSystemPath()
        } returns tempDir.toString()
    }
}
